### PR TITLE
Test cases for <code>AllocaInLoop</code>

### PR DESCRIPTION
**Pull request by zlaski**
Here are some test cases for <code>AllocaInLoop</code>.  The query is currently testing for <code>__builtin_alloca</code>; Visual C++ uses <code>_alloca</code> and <code>_malloca</code> so so we might want to check for those as well.

### CODE REVIEW

File: cpp/ql/src/Likely Bugs/Memory Management/AllocaInLoop.ql [hunk 1]
```diff
@@ -24,4 +24,4 @@ from Loop l, FunctionCall fc
 where getAnEnclosingLoopOfExpr(fc) = l
   and fc.getTarget().getName() = "__builtin_alloca"
   and not l.(DoStmt).getCondition().getValue() = "0"
-select fc, "Stack allocation is inside a $@ and could lead to overflow.", l, l.toString()
+select fc, "Stack allocation is inside a $@ and could lead to stack overflow; consider hoisting the allocation out of the loop, or using 'malloc()' or 'new' instead.", l, l.toString()
```
[the remark this answers is not in the document]
Comment (jbj) [reply cpp/ql/src/Likely Bugs/Memory Management/AllocaInLoop.ql]: That's right. See https://github.com/Semmle/ql/blob/master/docs/query-metadata-style-guide.md. I know we recently merged #1080, which is in contradiction of this policy, but generally we just want the alert message to point out what's wrong and not how to fix it.

**Comment by zlaski-semmle** [comment]
Any thoughts on including Microsoft's "_alloca" in the tests?  Our Seattle crew should dig it :-)

**Comment by rdmarsh2** [comment]
It would be good to get the Microsoft versions added to the tests in this PR. You could also add support for them as a separate commit in this PR if you'd like, or I can do it as a followup PR.

**Comment by zlaski-semmle** [comment]
I think it totally makes sense to do it in this PR since we're dealing with the same thing conceptually.

One wrinkle/challenge will be to modify the QL query accordingly.  It is not sufficient to look for "_alloca" (and perhaps "_malloca"); we must also know that the declaration originated in the <malloc.h> header.  This will be a good exercise for me, though I may ask for assistance.

**Comment by zlaski-semmle** [comment]
So here is an updated query designed to catch the Microsoftisms:

<pre>
from Loop l, FunctionCall fc
where
  getAnEnclosingLoopOfExpr(fc) = l and
  (
    fc.getTarget().getName() = "__builtin_alloca"
    or
    (
      (fc.getTarget().getName() = "_alloca" or fc.getTarget().getName() = "_malloca") and
      fc.getTarget().getADeclarationEntry().getFile().getBaseName() = "malloc.h"
    )
  ) and
  not l.(DoStmt).getCondition().getValue() = "0"
select fc, "Stack allocation is inside a $@ and could lead to stack overflow.", l, l.toString()
</pre>

It does compile in the LGTM console window, but I have yet to run it.  Any thoughts/corrections?

**Comment by zlaski-semmle** [comment]
I've updated the test cases and the query and all seems to work as expected (both the negative and positive cases).  The results are in a new PR:  https://github.com/Semmle/ql/pull/1107.

**Comment by jbj** [comment]
Superseded by #1107.

**Pull request closed without merging**